File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        if (!preorder.size())
            return nullptr;
        TreeNode* root = new TreeNode(preorder[0]);
        TreeNode* curr = root;
        int i = 0;
        return dfs(i, preorder, INT_MIN, INT_MAX);
    }

    TreeNode* dfs(int& i, vector<int>& preorder, int l, int r) {
        int n = preorder.size();
        if (i >= n)
            return nullptr;

        if(preorder[i]<=l || preorder[i]>=r) return nullptr;    

        TreeNode* root = new TreeNode(preorder[i++]);

        root->left = dfs(i, preorder, l, root->val);

        root->right = dfs(i, preorder, root->val, r);

        return root;
    }
};

File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private TreeNode construct(TreeNode root,int val){
        if(root==null)return new TreeNode(val);
        if(root.val>val){
            root.left=construct(root.left,val);
        }else{
            root.right=construct(root.right,val);
        }
        return root;
    }
    public TreeNode bstFromPreorder(int[] preorder) {
        int n=preorder.length;
        if(n==0)return null;
        TreeNode root=new TreeNode(preorder[0]);
        for(int i=1;i<n;i++){
            construct(root,preorder[i]);
        }
        return root;
    }
}